contextutil: RunWithTimeout preserves cause

jordanlewis opened the pull request:
Previously, RunWithTimeout could throw away the error cause in some
cases. Prevent this from happening.

Release note: None

jordanlewis committed:
contextutil: RunWithTimeout preserves cause

Previously, RunWithTimeout could throw away the error cause in some
cases. Prevent this from happening.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/35180)
<!-- Reviewable:end -->

andreimatei reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35180)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @jordanlewis)

___
*[pkg/util/contextutil/context.go, line 94 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35180#-LZ_xYejEsvZV221uBlK:-LZ_xYejEsvZV221uBlL:b-foqy4x) ([raw file](https://github.com/cockroachdb/cockroach/blob/3df5e9757f0c6bc8db2dcdbdf45114a6bf9f308a/pkg/util/contextutil/context.go#L94)):*
> ```Go
> 			operation: op,
> 			duration:  timeout,
> 			cause:     err,
> ```

I don't know... Ideally we'd overwrite the `DeadlineExceeded` at the bottom of the causes with a `TimeoutError` rather than insert a new error at the top. Except I don't think we can overwrite the bottom in the general case cause causes in `pkg.errors` errors are not mutable.

Inserting something at the top is also potentially incorrect. If an operation returns an error that has nothing to do with the deadline but the deadline has also elapsed, you don't want to say anything about deadline cause it'd be misleading: it would lead you to believe that the operation was canceled when, in fact, it failed for an unrelated reason.
So I'd rather we don't do anything on `ctx.Err() == context.DeadlineExceeded`, and instead only do something if we find a `DeadlineExceeded` in the cause hierarchy. Except as I was saying in the first paragraph, I'm not sure that changing the top error for a wrapped error is a good thing... So I guess I'd rather just do something when `err == context.DeadlineExceeded`. I don't think the TimeoutError adds much to other wrapped errors.

In general I guess I'd rather have this helper do as little as possible. In fact, I'd rather we get rid of it all-together and move to wrapping more errors in lower layers so that a naked `ContextDeadlineExceeded` doesn't get to SQL so often.


<!-- Sent from Reviewable.io -->

jordanlewis commented:
You're right that this is subtle to get right. It's really an annoying problem. Just changing this to do anything on raw `DeadlineExceeded` is really unsatisfying though because people constantly wrap random errors with a slightly more descriptive one, so then you'd end up just in the same situation we are in now, with something like `pq: transaction failed: context deadline exceeded` or whatever.

jordanlewis commented:
PTAL, removed the second case and added some tests. It'll still be confusing if somebody gets a timeout error that wasn't directly caused by the fresh context ,but I think that's a rare case.

andreimatei reviewed (approved):
LGTM

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35180#-:-LZfSDc10Aeck4Gvs14E:bb6ywmx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis)



<!-- Sent from Reviewable.io -->

jordanlewis commented:
bors r+

jordanlewis commented:
bors r+
